### PR TITLE
Clientside stateless resets

### PR DESCRIPTION
**Pull request by Ralith**
This adds support for clients performing stateless resets by transmitting `NEW_CONNECTION_ID` immediately after handshake to establish a reset token, and fixes several bugs encountered along the way.

### CODE REVIEW

**Comment by Ralith** [comment]
Managed to factor away the monolithic commit after all.

**Comment by Ralith** [comment]
Rebased. 72d5a80736c10ae944e4a8a186b990731f5e08a7 introduces a shared mutable hash table that provides the synchronization formerly implemented by indirection through `Multiplexer`, in a manner that supports eventual `Send` connection tasks. Synchronization through shared memory is useful here because connections may remove local connection IDs at any time due to remote prompting and may in principle issue new connection IDs at any time, while the dispatch table used by the `Endpoint` must be kept up to date. Overhead is expected to be low as both of these events should be rare.

An alternative option that avoids the `Mutex` would be to have all issuings of new connection IDs initiated by the `Endpoint`, which would be sufficient for the main usecases of client-side stateless resets and of avoiding linkability across migrations. Retiring old CIDs could be done via I/O messages, similar to timer updates.

**Comment by Ralith** [comment]
> Is there value in keeping the other fields outside the `Mutex`?

On review, there's no functional benefit to this. However, I do think it's nice to limit the scope of interior mutability of any kind as narrowly as possible.

**Comment by Ralith** [comment]
> Maybe it's time to bite the bullet and switch to channels for Endpoint <-> Connection communication?

This is onerous if `Connection`s issue new CIDs on their own initiative, because it would involve requesting one and waiting for a new response. I can draft the alternative approach I outlined in a prior comment instead, removing that capability in favor of preallocating (and perhaps eventually dynamically issuing) additional CIDs in the `Endpoint`, with retirement mediated by `Io` messages.

In general, I don't think there will be need to make explicit use of channels inside `Connection` when we can expose methods that consume/yield events, which can be wired up however the caller likes.

**Comment by djc** [comment]
In my mind connections would be fully responsible for creating/issuing/retring CIDs, merely informing their `Endpoint` of (changes to) the current set of allowed CIDs that may reference that `Connection`.

I don't really see how we can avoid channels when we want to run both `Endpoint`s and `Connection`s in separate tasks?

**Comment by Ralith** [comment]
> In my mind connections would be fully responsible for creating/issuing/retring CIDs, merely informing their Endpoint of (changes to) the current set of allowed CIDs that may reference that Connection.

On review, the draft recommends that peers be supplied with at least eight CIDs at all times, so I think this can be much simpler: the `Endpoint` supply every new connection with eight CIDs to issue when established, and whenever `conn.poll_io() == Io::RetireCid(_)`, the `Endpoint` will supply a new CID to replace the retired one. No further coordination should be necessary.

> I don't really see how we can avoid channels when we want to run both Endpoints and Connections in separate tasks?

Channels are necessary for that, but the `Connection` doesn't have to know about them. My thinking is that `Connection` retains its pure, synchronous interface, and may (or may not) be driven by asynchronous communication with an `Endpoint` as needed. This also makes it easy to start testing `Connection` on its own, which seems nice.

**Comment by djc** [comment]
Okay, that sounds pretty good.

**Comment by Ralith** [comment]
Drafted.

File: quinn-proto/src/endpoint.rs [hunk 1]
```diff
@@ -379,12 +379,21 @@ impl Endpoint {
...
+            let remaining = self.connections[conn_id.0].handle_decode(now, ecn, partial_decode);
+            if was_handshake && !self.connections[conn_id.0].is_handshaking() {
+                // Newly established connection
+                if self.connections[conn_id.0].side().is_server() {
+                    self.incoming_handshakes -= 1;
+                    self.ctx.incoming.push_back(conn_id);
+                }
+                if self.config.local_cid_len != 0 {
+                    // endpoints SHOULD provide and maintain at least eight connection IDs
+                    for _ in 0..8 {
```
Comment (djc) [inline quinn-proto/src/endpoint.rs]: Should we somehow document what this `8` is?

File: quinn-proto/src/frame.rs [hunk 2]
```diff
@@ -120,42 +120,21 @@ pub enum Frame {
...
-        id: StreamId,
-    },
-    StopSending {
-        id: StreamId,
-        error_code: u16,
-    },
-    RetireConnectionId {
-        sequence: u64,
-    },
+    Blocked { offset: u64 },
```
Comment (djc) [inline quinn-proto/src/frame.rs]: Unrelated style changes?

File: quinn-proto/src/frame.rs [hunk 3]
```diff
@@ -569,11 +548,11 @@ impl Iter {
                 }
                 let mut reset_token = [0; RESET_TOKEN_SIZE];
                 self.bytes.copy_to_slice(&mut reset_token);
-                Frame::NewConnectionId {
+                Frame::NewConnectionId(NewConnectionId {
```
Comment (djc) [inline quinn-proto/src/frame.rs]: Separate commit for this type extraction?

File: quinn-proto/src/endpoint.rs (hunk 1 above)
Comment (djc) [reply quinn-proto/src/endpoint.rs]: Also it feels odd for the majority of this hunk to live in `Endpoint`, why is that?

File: quinn-proto/src/connection.rs [hunk 4]
```diff
@@ -2798,7 +2812,13 @@ pub enum Io {
         packet: Box<[u8]>,
     },
     /// Stop or (re)start a timer
-    TimerUpdate { timer: Timer, update: TimerUpdate },
+    TimerUpdate {
```
Comment (djc) [inline quinn-proto/src/connection.rs]: Unrelated style change.

File: quinn-proto/src/endpoint.rs (hunk 1 above)
Comment (Ralith) [reply quinn-proto/src/endpoint.rs]: Most of the changes in this hunk are semantic no-ops to ensure we can call `self.new_cid()` without an outstanding borrow. That said, I think the placement is justified. We're doing two things here: reporting the establishment of a new incoming connection, and providing new connections with their alternate CIDs. Both of these inherently involve modifying `Endpoint` state, for the incoming connection queue and the connection ID table respectively.

I'll insert a constant for the CID count.

File: quinn-proto/src/frame.rs (hunk 2 above)
Comment (Ralith) [reply quinn-proto/src/frame.rs]: rustfmt does this directly in response to the type extraction, for some reason.

File: quinn-proto/src/connection.rs (hunk 4 above)
Comment (Ralith) [reply quinn-proto/src/connection.rs]: rustfmt does this directly in response to the added enum variant, for some reason.

**Comment by Ralith** [comment]
I think that covers all the comments.

**Comment by Ralith** [comment]
Rebased.
